state/relationunit.go

```diff
@@ -244,17 +244,31 @@ func (ru *RelationUnit) subordinateOps() ([]txn.Op, string, error) {
 		return nil, "", err
 	}
 	if len(related) != 1 {
-		return nil, "", fmt.Errorf("expected single related endpoint, got %v", related)
+		return nil, "", errors.Errorf("expected single related endpoint, got %v", related)
 	}
+	// Find the machine ID that the principal unit is deployed on, and use
+	// that for the subordinate. It is worthwhile noting that if the unit is
+	// in a CAAS model, there are no machines.
+	principal, err := ru.st.Unit(ru.unitName)
+	if err != nil {
+		return nil, "", errors.Annotate(err, "unable to load principal unit")
+	}
+	// We dont care just now if the machine isn't assigned, as CAAS models
+	// will return that error. For IAAS models, the machine *should* always
+	// be assigned before it is able to enter scope.
+	principalMachineID, _ := principal.AssignedMachineId()
```

Comment:
It's dangerous to ignore all errors as the implementation might change and introduce a bug.
It is safer to use the ShouldBeAssigned() helper. This would also help document all cases where we make a caas vs iaas decision to help the refactoring we'd ideally like to do later.

Reply:
Added a ShouldBeAssigned check, but still ignore the error because of the amount of test fallout.